sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala

```diff
@@ -83,10 +88,221 @@ object ReorderJoin extends Rule[LogicalPlan] with PredicateHelper {
+   * represents Project or Filter operators above a LeafNode. Conservatively, the algorithm
+   * only considers base table access as part of a star join since they provide reliable
+   * statistics.
+   *
+   * If there are no eligible fact tables (e.g. no base table access i.e. complex sub-plans),
+   * the algorithm falls back to the positional join reordering, which is the default join
+   * ordering in Catalyst. Otherwise, the algorithm finds the largest fact table by sorting
+   * the tables based on their size
+   *
+   * The algorithm next computes the set of dimension tables for the current fact table.
+   * Conservatively, only equality joins are considered between a fact and a dimension table.
+   *
+   * Given a star join, i.e. fact and dimension tables, the algorithm considers three cases:
+   *
+   * 1) The star join is an expanding join i.e. the fact table is joined using inequality
+   * predicates, or Cartesian product. In this case, the algorithm is looking for the next
+   * eligible star join. The rational is that a selective star join, if found, may reduce
+   * the data stream early in the plan. An alternative, more conservative design, is to simply
+   * fall back to the default join reordering if the star join with the largest fact table
+   * is not selective.
+   *
+   * 2) The star join is a selective join. This case is detected by observing local predicates
+   * on the dimension tables. In a star schema relationship, the join between the fact and the
+   * dimension table is in general a FK-PK join. Heuristically, a selective dimension may reduce
+   * the result of a join.
+   *
+   * 3) The star join is not a selective join (i.e. doesn't reduce the number of rows on the
+   * driving arm). In this case, the algorithm conservatively falls back to the default join
+   * reordering.
+   */
+  @tailrec
+  private def findEligibleStarJoinPlan(
+      joinedTables: Seq[(LogicalPlan, InnerLike)],
+      eligibleFactTables: Seq[(LogicalPlan, InnerLike)],
+      conditions: Seq[Expression]): Seq[(LogicalPlan, InnerLike)] = {
+    if (eligibleFactTables.isEmpty || joinedTables.size < 2) {
+      // There is no eligible star join.
+      Seq.empty[(LogicalPlan, InnerLike)]
+    } else {
+      // Compute the size of the eligible fact tables and sort them in descending order.
+      // An eligible fact table is a base table access with valid statistics.
+      val sortedFactTables = eligibleFactTables.map { plan =>
+        TableSize(plan, computeBaseTableSize(plan._1, conditions))
+      }.collect {
+        case t @ TableSize(_, Some(_)) => t
+      }.sortBy(_.size)(implicitly[Ordering[Option[BigInt]]].reverse)
+
+      sortedFactTables match {
+        case Nil =>
+          // There are no stats available for these plans.
+          // Return an empty plan list and fall back to the default join reordering.
+          Seq.empty[(LogicalPlan, InnerLike)]
+
+        case table1 :: table2 :: _ if table1.size == table2.size =>
```

Comment:
The “size” should represent table cardinality after applying the pushed down local predicates  i.e. num_rows*selectivity. Temporarily, I used the sizeinBytes value since the join strategies are also using this value to make planning decisions. Long term, a fact table will be determined based on the referential integrity constraints with the other tables. Then, the star join will be planned based on joins’ selectivity.